File: backend/security.py
"""
Security utilities for authentication and password handling.
Includes JWT token creation, password hashing, and user authentication.
"""
from datetime import datetime, timedelta
from typing import Optional
from fastapi import Depends, HTTPException, status
from fastapi.security import OAuth2PasswordBearer
from passlib.context import CryptContext
from jose import JWTError, jwt

from config import SECRET_KEY, ALGORITHM, ACCESS_TOKEN_EXPIRE_MINUTES
from database import get_db_connection
from models import User, TokenData

# Password hashing context
pwd_context = CryptContext(schemes=["bcrypt"], deprecated="auto")

# OAuth2 scheme for token authentication
oauth2_scheme = OAuth2PasswordBearer(tokenUrl="auth/login")


def verify_password(plain_password: str, hashed_password: str) -> bool:
    """Verify a plain password against a hashed password."""
    return pwd_context.verify(plain_password, hashed_password)


def get_password_hash(password: str) -> str:
    """Hash a password using bcrypt."""
    return pwd_context.hash(password)


def create_access_token(data: dict, expires_delta: Optional[timedelta] = None) -> str:
    """
    Create a JWT access token.
    
    Args:
        data: Dictionary containing token payload (typically {"sub": user_id})
        expires_delta: Optional custom expiration time
    
    Returns:
        Encoded JWT token string
    """
    to_encode = data.copy()
    if expires_delta:
        expire = datetime.utcnow() + expires_delta
    else:
        expire = datetime.utcnow() + timedelta(minutes=ACCESS_TOKEN_EXPIRE_MINUTES)
    to_encode.update({"exp": expire})
    
    # Ensure sub is string for JWT standard
    if "sub" in to_encode:
        to_encode["sub"] = str(to_encode["sub"])
    
    encoded_jwt = jwt.encode(to_encode, SECRET_KEY, algorithm=ALGORITHM)
    return encoded_jwt


async def get_current_user(token: str = Depends(oauth2_scheme)) -> User:
    """
    Dependency to get the current authenticated user from JWT token.
    
    Args:
        token: JWT token from Authorization header
    
    Returns:
        User object with user details
    
    Raises:
        HTTPException: If token is invalid or user not found
    """
    credentials_exception = HTTPException(
        status_code=status.HTTP_401_UNAUTHORIZED,
        detail="Could not validate credentials",
        headers={"WWW-Authenticate": "Bearer"},
    )
    
    try:
        payload = jwt.decode(token, SECRET_KEY, algorithms=[ALGORITHM])
        user_id_str: str = payload.get("sub")
        if user_id_str is None:
            raise credentials_exception
        
        # Convert string back to int
        user_id = int(user_id_str)
        token_data = TokenData(user_id=user_id)
    except (JWTError, ValueError, TypeError) as e:
        print(f"Token validation error: {e}")
        raise credentials_exception
    
    # Fetch user from database
    conn = get_db_connection()
    cursor = conn.cursor()
    try:
        cursor.execute(
            "SELECT id, email, name, created_at FROM users WHERE id = %s",
            (token_data.user_id,)
        )
        user_row = cursor.fetchone()
        
        if user_row is None:
            raise credentials_exception
        
        return User(
            id=user_row[0],
            email=user_row[1],
            name=user_row[2],
            created_at=user_row[3]
        )
    finally:
        conn.close()

File: backend-node/src/security.ts
import bcrypt from 'bcrypt';
import jwt from 'jsonwebtoken';
import { Request, Response, NextFunction } from 'express';
import { getDbPool } from './database';

const SECRET_KEY = process.env.SECRET_KEY || "09d25e094faa6ca2556c818166b7a9563b93f7099f6f0f4caa6cf63b88e8d3e7";
const ALGORITHM = "HS256";
const ACCESS_TOKEN_EXPIRE_MINUTES = parseInt(process.env.ACCESS_TOKEN_EXPIRE_MINUTES || "30", 10);

export async function verifyPassword(plainPassword: string, hashedPassword: string): Promise<boolean> {
    return bcrypt.compare(plainPassword, hashedPassword);
}

export async function getPasswordHash(password: string): Promise<string> {
    const saltRounds = 12; // passlib auto default equivalent
    return bcrypt.hash(password, saltRounds);
}

export function createAccessToken(data: Record<string, any>, expiresInMinutes?: number): string {
    const payload = { ...data };
    
    if (payload.sub !== undefined) {
        payload.sub = String(payload.sub);
    }
    
    const minutes = expiresInMinutes || ACCESS_TOKEN_EXPIRE_MINUTES;
    
    return jwt.sign(payload, SECRET_KEY, {
        algorithm: ALGORITHM as jwt.Algorithm,
        expiresIn: `${minutes}m`
    });
}

// User type definition
export interface User {
    id: number;
    email: string;
    name: string;
    created_at: Date;
}

// Express Request extension for authenticated routes
export interface AuthRequest extends Request {
    user?: User;
}

export async function getCurrentUser(req: AuthRequest, res: Response, next: NextFunction) {
    const authHeader = req.headers.authorization;
    
    if (!authHeader || !authHeader.startsWith('Bearer ')) {
        return res.status(401).json({ detail: "Not authenticated" });
    }
    
    const token = authHeader.split(' ')[1];
    
    try {
        const decoded = jwt.verify(token, SECRET_KEY) as jwt.JwtPayload;
        
        const userIdStr = decoded.sub;
        if (!userIdStr) {
            return res.status(401).json({ detail: "Could not validate credentials" });
        }
        
        const userId = parseInt(userIdStr, 10);
        
        const pool = getDbPool();
        const result = await pool.query(
            "SELECT id, email, name, created_at FROM users WHERE id = $1",
            [userId]
        );
        
        if (result.rows.length === 0) {
            return res.status(401).json({ detail: "Could not validate credentials" });
        }
        
        const userRow = result.rows[0];
        req.user = {
            id: userRow.id,
            email: userRow.email,
            name: userRow.name,
            created_at: userRow.created_at
        };
        
        next();
    } catch (e) {
        console.error(`Token validation error: ${e}`);
        return res.status(401).json({ detail: "Could not validate credentials" });
    }
}
